plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcClient.java

```diff
@@ -89,13 +90,25 @@ PreparedQuery prepareQuery(
             List<JdbcColumnHandle> columns,
             Map<String, String> columnExpressions);
 
+    default boolean supportsTopNPushdown(ConnectorSession session, JdbcTableHandle handle, Map<String, ColumnHandle> assignments, List<SortItem> sortOrder)
+    {
+        return false;
+    }
+
+    default PreparedQuery implementTopNQuery(PreparedQuery preparedQuery, List<SortItem> sortItems)
```

Comment:
Since this is not called from outside, it doesn't need to be in JdbcClient interface.

I think the reason we have it here, is because I wanted to keep sort items out of the JdbcTableHandle, and so the logic resided in JdbcMetadata and needed to call JdbcClient for help.

Now, the state is moved (moved back) to the table handle and being handled in prepareQuery / buildSql methods. But JdbcMetadata still needs to know whether TopN pushdown is supported. For this reason I think we need to model this like we do with limit (and as you had this originally)

- interface JdbcClient.supportsTopN
- interface JdbcClient.isTopNLimitGuaranteed
- implementation BaseJdbcClient.supportsTopN inspecting `topNFunction().isPresent()`
- protected BaseJdbcClient.topNFunction returning empty by default